ostd/src/cpu/local/dyn_cpu_local.rs

```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Dynamically-allocated CPU-local objects.
+
+use alloc::vec::Vec;
+use core::marker::PhantomData;
+
+use align_ext::AlignExt;
+use bitvec::prelude::{bitvec, BitVec};
+
+use super::{AnyStorage, CpuLocal};
+use crate::{
+    cpu::{all_cpus, num_cpus, CpuId, PinCurrentCpu},
+    mm::{paddr_to_vaddr, FrameAllocOptions, Segment, Vaddr, PAGE_SIZE},
+    sync::SpinLock,
+    trap::DisabledLocalIrqGuard,
+    Result,
+};
+
+/// A dynamically-allocated storage for a CPU-local variable of type `T`.
+///
+/// Such a CPU-local storage is not intended to be allocated directly.
+/// Use `osdk_heap_allocator::alloc_cpu_local` instead.
+pub struct DynamicStorage<T>(*const T);
+
+unsafe impl<T> AnyStorage<T> for DynamicStorage<T> {
+    fn get_ptr_on_current(&self, guard: &DisabledLocalIrqGuard) -> *const T {
+        self.get_ptr_on_target(guard.current_cpu())
+    }
+
+    fn get_ptr_on_target(&self, cpu_id: CpuId) -> *const T {
+        let bsp_va = self.0 as usize;
+        let va = bsp_va + cpu_id.as_usize() * CHUNK_SIZE;
+        va as *const T
+    }
+}
+
+impl<T> CpuLocal<T, DynamicStorage<T>> {
+    /// Creates a new dynamically-allocated CPU-local object, and
+    /// initializes it with `init_values`.
+    ///
+    /// The given `ptr` points to the variable located on the BSP.
+    /// The corresponding variables on all CPUs are initialized to zero.
+    ///
+    /// Please do not call this function directly. Instead, use
+    /// `osdk_heap_allocator::alloc_cpu_local`.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that the new per-CPU object belongs to an
+    /// existing `DynCpuLocalChunk`, and does not overlap with any existing
```

Comment:
You can use Rust [intra-doc links](https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html) for such references. I would write it as:

```suggestion
    /// existing [`DynCpuLocalChunk`], and does not overlap with any existing
```